js/modules/loki_file_server_api.js

```diff
@@ -43,6 +53,182 @@ class LokiFileServerAPI {
     return this._setOurDeviceMapping(authorisations, isPrimary);
   }
 
+  async getDeviceMappingForUsers(pubKeys) {
+    const users = await this._server.getUsers(pubKeys);
+    return users;
+  }
+
+  async verifyUserObjectDeviceMap(pubKeys, isRequest, iterator) {
+    const users = await this.getDeviceMappingForUsers(pubKeys);
+    // log.info('verifyUserObjectDeviceMap Found', users.length, 'users')
+
+    // go through each user and find deviceMap annotations
+    const notFoundUsers = [];
+    users.forEach(user => {
+      let found = false;
+      if (!user.annotations || !user.annotations.length) {
+        log.info(
+          `verifyUserObjectDeviceMap no annotation for ${user.username}`
+        );
+        return;
+      }
+      user.annotations.forEach(note => {
+        if (note.type !== 'network.loki.messenger.devicemapping') {
+          return;
+        }
+        // isn't desired type
+        // request is slave => primary type...
+        if (
+          (isRequest && note.value.isPrimary !== '0') ||
+          (!isRequest && note.value.isPrimary === '0')
+        ) {
+          /* log.info(`verifyUserObjectDeviceMap found wrong type of` +
+                  `relationship ${user.username}`); */
+          // console.log(`https://file.lokinet.org/users/@${user.username}?prettyPrint=1&include_annotations=1`);
+          return;
+        }
+        const { authorisations } = note.value;
+        if (!Array.isArray(authorisations)) {
+          return;
+        }
+        authorisations.forEach(auth => {
+          // log.info('devmap auth', auth);
+          // only skip, if in secondary search mode
+          if (isRequest && auth.secondaryDevicePubKey !== user.username) {
+            // this is not the authorization we're looking for
+            log.info(
+              `Request and ${auth.secondaryDevicePubKey} != ${user.username}`
+            );
+            return;
+          }
+          // log.info('auth', auth);
+          try {
+            // request (secondary wants to be paired with this primary)
+            // grant (primary approves this secondary)
+            window.libloki.crypto.verifyPairingSignature(
+              auth.primaryDevicePubKey,
+              auth.secondaryDevicePubKey,
+              dcodeIO.ByteBuffer.wrap(
+                isRequest ? auth.requestSignature : auth.grantSignature,
+                'base64'
+              ).toArrayBuffer(),
+              isRequest
```

Comment:
I don't think this will work and I apologise for not being able to clear up the confusion with request/grant from the beginning.

So basically like we talked the other day, every authorisation contains 2 signatures which both need to be verified.
So if you unroll the loop you should have this instead:
```js
window.libloki.crypto.verifyPairingSignature(p1,p2, requestSignature, Type.REQUEST);
window.libloki.crypto.verifyPairingSignature(p1,p2, grantSignature, Type.GRANT);
```
So I think the variable `isRequest` should be replaced with `userIsSecondary` or something similar. But sincet the authorisation held by the primary device is the same as the authorisation held by the secondary device, the cryptographic verification of the signature should be independent from whether the user is a primary or a secondary device. Hit me on discord if it's still not clear?